### PR TITLE
deepin.startdde: init at 5.0.1

### PR DESCRIPTION
**Pull request by romildo**
###### Motivation for this change

- Add [startdde](https://github.com/linuxdeepin/startdde), used for launching DDE components and invoking user's custom applications which are compliant with xdg autostart specification.

- Add the nixos desktop service module.

The packaging of Deepin Desktop Environment is alpha quality at this point. But at least it should be possible to run DDE for testing and bug fix purposes.

About packaging deepin for NixOS: #59023

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by romildo** [comment]
Screenshot from 2019-06-26 11-14-47
Screenshot from 2019-06-26 11-15-25
Screenshot from 2019-06-26 11-19-41

**Review by worldofpeace — CHANGES_REQUESTED** [review]
I have a commit you should probably integrate here https://github.com/worldofpeace/nixpkgs/commit/53568a5dd5dfa216ca8a601373a77ef90fa83ad3.
This enables more services that are needed and uses an upstream session, etc.

Also think we should get a module for the greeter here while we're at it https://github.com/worldofpeace/nixpkgs/commit/8314f034590f4b818a5cb736e1a7e12fecce7318 (wip)

**Comment by romildo** [comment]
> I have a commit you should probably integrate here [worldofpeace@53568a5](https://github.com/worldofpeace/nixpkgs/commit/53568a5dd5dfa216ca8a601373a77ef90fa83ad3).
> This enables more services that are needed and uses an upstream session, etc.

Added.

> Also think we should get a module for the greeter here while we're at it [worldofpeace@8314f03](https://github.com/worldofpeace/nixpkgs/commit/8314f034590f4b818a5cb736e1a7e12fecce7318) (wip)

Added.

File: pkgs/desktops/deepin/startdde/default.nix
Comment (romildo) [reply pkgs/desktops/deepin/startdde/default.nix]: Done.

File: pkgs/desktops/deepin/startdde/default.nix
Comment (romildo) [reply pkgs/desktops/deepin/startdde/default.nix]: Done.

**Comment by worldofpeace** [comment]
Two other commits for you @romildo 

* https://github.com/worldofpeace/nixpkgs/commit/310f0764f6f02062d13478deec715af00975164d
fixes build

* https://github.com/worldofpeace/nixpkgs/commit/d1a7152c03ba8745479d2a66f798ea724f990957
integrate after #65248 is on master

Could you also rebase?

**Comment by worldofpeace** [comment]
ping @romildo.

I think we can integrate this after figuring out the default font names for deepin :crossed_fingers: 
After that I plan to write a simple test.

**Comment by romildo** [comment]
> Two other commits for you @romildo
> 
>     * [worldofpeace@310f076](https://github.com/worldofpeace/nixpkgs/commit/310f0764f6f02062d13478deec715af00975164d)
>       fixes build
> 
>     * [worldofpeace@d1a7152](https://github.com/worldofpeace/nixpkgs/commit/d1a7152c03ba8745479d2a66f798ea724f990957)
>       integrate after #65248 is on master
> 
> 
> Could you also rebase?

Done.

I had to set an older version for `vala` to build `deepin-wm` due to compilation errors with the latest version.

Sorry for the delay. I am in vacations (first time in Europe).

**Comment by worldofpeace** [comment]
> Sorry for the delay. I am in vacations (first time in Europe).

In that case can you make your branch writable to maintainers?

**Comment by romildo** [comment]
> > Sorry for the delay. I am in vacations (first time in Europe).
> 
> In that case can you make your branch writable to maintainers?

It is already marked as writable to maintainers.

**Comment by worldofpeace** [comment]
> > > Sorry for the delay. I am in vacations (first time in Europe).
> > 
> > 
> > In that case can you make your branch writable to maintainers?
> 
> It is already marked as writable to maintainers.

Weird, last time I tried to push it was rejected.
Guess it's good now though :+1:

**Comment by romildo** [comment]
> > > > Sorry for the delay. I am in vacations (first time in Europe).
> > > 
> > > 
> > > In that case can you make your branch writable to maintainers?
> > 
> > 
> > It is already marked as writable to maintainers.
> 
> Weird, last time I tried to push it was rejected.
> Guess it's good now though +1

See the screenshot. It is marked.
Screenshot from 2019-08-08 10-35-04

Or should I do something else?

**Comment by romildo** [comment]
> > > > Sorry for the delay. I am in vacations (first time in Europe).
> > > 
> > > 
> > > In that case can you make your branch writable to maintainers?
> > 
> > 
> > It is already marked as writable to maintainers.
> 
> Weird, last time I tried to push it was rejected.
> Guess it's good now though +1

It is marked here:

Screenshot from 2019-08-08 10-37-39

Should I do anything else?

**Comment by worldofpeace** [comment]
Yeah that's all you have to do @romildo, thanks.

**Comment by romildo** [comment]
Rebased.

**Comment by worldofpeace** [comment]
I noticed that I need to adjust the wrapping of these things more, and the addition of `wrapQtAppsHook` has added an interesting theming issue. We probably need to re-add a library to `systemPackages`.

I have stuff locally for the wrappers though.

**Comment by worldofpeace** [comment]
d080461209327a8660e0af8891264e1db52dc145..d64bb6d2d337338234d31f4ed9696c9140278bd6 fixed the aforementioned issue.

Though there is an issue with usability:
Screenshot from 2019-08-11 14 58 05

Certain applications don't get their respective icons in `dde-launcher`.
I'm not expecting that logic to be implemented in that software though (lower in deepin).

---

Other than this I think I'll give this a boot entry on my machine to test it further.

**Comment by worldofpeace** [comment]
I think it's possible the aforementioned issue is fixed with the new PAM env.
Additionally deepin has deprecated deepin-wm and now uses kwin. So any issues that bamf currently had will go away with the switch to kwin.

However deepin-wm is broken currently... and no fix in sight so I can't merge atm.
Will open a PR trying to package deepin's kwin though.

**Comment by worldofpeace** [comment]
Ok, so I did try this locally with dde-kwin and it just starts up with a white screen. Need to investigate further.

**Comment by romildo** [comment]
Resolved two conflicts with the base branch.

**Comment by worldofpeace** [comment]
I've moved the modules into https://github.com/NixOS/nixpkgs/pull/77293 so we can integrate this and get one small step closer. Will merge after eval succeeds.

Here's the diff of what was dropped from this PR
* https://gist.github.com/worldofpeace/723e197421073b046a5a0dd6c9b7bffb

and I had to rewrite two commits into https://github.com/NixOS/nixpkgs/pull/77293 so anything here didn't touch the modules.

**Pull request merged**
